deepinv/training_utils/trainer.py

```diff
@@ -0,0 +1,714 @@
+    ckpt_pretrained: Union[str, None] = None
+    freq_plot: int = 1
+    verbose_individual_losses: bool = True
+    display_losses_eval: bool = False
+
+    def setup_train(self):
+        r"""
+        Set up the training process.
+
+        It initializes the wandb logging, the different metrics, the save path, the physics and dataloaders,
+        and the pretrained checkpoint if given.
+        """
+
+        if type(self.train_dataloader) is not list:
+            self.train_dataloader = [self.train_dataloader]
+
+        if self.eval_dataloader and type(self.eval_dataloader) is not list:
+            self.eval_dataloader = [self.eval_dataloader]
+
+        self.save_path = Path(self.save_path)
+
+        self.G = len(self.train_dataloader)
+
+        if self.wandb_setup is not None and not self.wandb_vis:
+            warnings.warn(
+                "wandb_vis is False but wandb_setup is provided. Wandb visualization deactivated (wandb_vis=False)."
+            )
+
+        # wandb initialization
+        if self.wandb_vis:
+            if wandb.run is None:
+                wandb.init(**self.wandb_setup)
+
+        if not isinstance(self.losses, list) or isinstance(self.losses, tuple):
+            self.losses = [self.losses]
+
+        if not isinstance(self.metrics, list) or isinstance(self.metrics, tuple):
+            self.metrics = [self.metrics]
+
+        # losses
+        self.logs_total_loss_train = AverageMeter("Training loss", ":.2e")
+        self.logs_losses_train = [
+            AverageMeter("Training loss " + l.name, ":.2e") for l in self.losses
+        ]
+
+        self.logs_total_loss_eval = AverageMeter("Validation loss", ":.2e")
+        self.logs_losses_eval = [
+            AverageMeter("Validation loss " + l.name, ":.2e") for l in self.losses
+        ]
+
+        # metrics
+        self.logs_metrics_train = [
+            AverageMeter("Training metric " + l.__class__.__name__, ":.2e")
+            for l in self.metrics
+        ]
+
+        self.logs_metrics_eval = [
+            AverageMeter("Validation metric " + l.__class__.__name__, ":.2e")
+            for l in self.metrics
+        ]
+
+        # gradient clipping
+        if self.check_grad:
+            self.check_grad_val = AverageMeter("Gradient norm", ":.2e")
+
+        self.save_path = f"{self.save_path}/{get_timestamp()}"
+
+        # count the overall training parameters
+        params = sum(p.numel() for p in self.model.parameters() if p.requires_grad)
+        print(f"The model has {params} trainable parameters")
+
+        # make physics and data_loaders of list type
+        if type(self.physics) is not list:
+            self.physics = [self.physics]
+
+        self.loss_history = []
+
+        self.epoch_start = 0
+        if self.ckpt_pretrained is not None:
+            checkpoint = torch.load(self.ckpt_pretrained)
+            self.optimizer.load_state_dict(checkpoint["optimizer"])
+            self.epoch_start = checkpoint["epoch"]
```

Comment:
Where has `self.model.load_state_dict(checkpoint["state_dict"])` gone?

Reply:
good point, this was removed by mistake